### PR TITLE
awsutil: ensure GenerateCredentialChain checks envs

### PR DESCRIPTION
**Pull request by fairclothjm**
https://github.com/hashicorp/go-secure-stdlib/pull/57 caused a regression in Vault. In that PR the AWS environment variable checks in GenerateCredentialChain were moved to NewCredentialsConfig for AWS_ROLE_ARN, AWS_WEB_IDENTITY_TOKEN_FILE, and AWS_ROLE_SESSION_NAME. This caused users of the awsutil library that call GenerateCredentialConfig to fail to set the above values if they are depending on env variables to be read.

Another solution could be to find all users of go-secure-stdlib’s awsutil and ensure env vars are properly read and set in the config before calling GenerateCredentialConfig like in https://github.com/hashicorp/vault/pull/21930. However, I think a this PR is a better resolution. With this change we shouldn't have to make any changes anywhere else.

This change was tested with the following steps:
- Create an AWS EKS cluster
- [Connect the EKS cluster as an OIDC provider for AWS IAM](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html)
- Create an AWS KMS key
- [Create an AWS IAM role with an assume-role policy authorizing that OIDC provider](https://docs.aws.amazon.com/eks/latest/userguide/associate-service-account-role.html), and an inline policy authorizing that IAM role for actions on the KMS key you created
- Deploy Vault via Vault Helm version 0.24.1 with an "awskms" seal, setting only kms_key_id, and adding a ServiceAccount annotation eks.amazonaws.com/role-arn: <IAM role ARN>
- Upgrade Vault via Vault Helm with a Vault image that includes this fix via `go get github.com/hashicorp/go-secure-stdlib/awsutil@d18ccdf3e9fb`

Additionally, this has been tested against the Boundary E2E tests by @ddebko. Thanks!

Closes https://github.com/hashicorp/go-secure-stdlib/issues/71

### CODE REVIEW

**Review by swenson — APPROVED** [review]
LGTM -- might be nice to add some unit tests though so we don't get another regression? This is a pretty long function with lots of edge cases.

File: awsutil/generate_credentials.go [hunk 1]
```diff
@@ -240,19 +251,19 @@ func (c *CredentialsConfig) GenerateCredentialChain(opt ...Option) (*credentials
...
 			}
 		} else if c.WebIdentityToken != "" {
-			c.log(hclog.Debug, "adding web identity provider with token", "roleARN", c.RoleARN)
+			c.log(hclog.Debug, "adding web identity provider with token", "roleARN", roleARN)
 			sess, err := session.NewSession()
 			if err != nil {
 				return nil, errors.Wrap(err, "error creating a new session to create a WebIdentityRoleProvider with token")
 			}
-			webIdentityProvider := stscreds.NewWebIdentityRoleProviderWithToken(sts.New(sess), c.RoleARN, c.RoleSessionName, FetchTokenContents(c.WebIdentityToken))
+			webIdentityProvider := stscreds.NewWebIdentityRoleProviderWithToken(sts.New(sess), roleARN, roleSessionName, FetchTokenContents(c.WebIdentityToken))
```
Comment (tomhjp) [inline awsutil/generate_credentials.go:266]: Was this one missed on purpose?

```suggestion
			webIdentityProvider := stscreds.NewWebIdentityRoleProviderWithToken(sts.New(sess), roleARN, roleSessionName, FetchTokenContents(tokenPath))
```

**Review by tomhjp — COMMENTED** [review]
LGTM, just one query

File: awsutil/generate_credentials.go (hunk 1 above)
Comment (fairclothjm) [reply awsutil/generate_credentials.go:266]: I don't believe this is correct. `WebIdentityToken` is the token contents not the file path. Whereas `tokenPath` is the `WebIdentityTokenFile`

https://github.com/hashicorp/go-secure-stdlib/blob/d18ccdf3e9fbe14614481f54e40380b468502cf6/awsutil/generate_credentials.go#L226

**Comment by fairclothjm** [comment]
> might be nice to add some unit tests though so we don't get another regression? This is a pretty long function with lots of edge cases.

@swenson Agreed! I will address adding some tests in another PR if you are ok with that?

**Review by maxcoulombe — APPROVED** [review]

File: awsutil/generate_credentials.go (hunk 1 above)
Comment (tomhjp) [reply awsutil/generate_credentials.go:266]: Oops, sorry! SGTM

**Review by tomhjp — APPROVED** [review]
